File: templates/tool-repo-template/examples/advanced.ts
/**
 * Advanced usage patterns for tool-name
 *
 * This example demonstrates:
 * - Performance optimization patterns
 * - Resource limit handling
 * - Batch processing with pre-allocation
 * - Streaming patterns for large data
 * - Error handling strategies
 *
 * Run this example:
 *   npx tsx examples/advanced.ts
 */

import { process, type Config, type Result } from '../src/index.js';

/**
 * Example: Pre-allocate arrays when size is known
 *
 * Performance pattern:
 * - new Array(size) pre-allocates memory
 * - Avoids dynamic resizing and copying
 * - Improves performance for known-size collections
 */
function batchProcessOptimized(inputs: string[], config: Config): Result[] {
  // Pre-allocate: we know we'll have exactly inputs.length results
  const results = new Array<Result>(inputs.length);

  for (let i = 0; i < inputs.length; i++) {
    results[i] = process(inputs[i], config);
  }

  return results;
}

/**
 * Example: Resource limit pattern
 *
 * Pattern used in tools that handle large inputs:
 * - Define constants for resource limits
 * - Check limits before processing
 * - Provide clear error messages
 */
const MAX_INPUT_LENGTH = 1_000_000; // 1 million characters

function processWithLimit(input: string, config: Config): Result {
  // Check resource limit before processing
  if (input.length > MAX_INPUT_LENGTH) {
    return {
      success: false,
      data: '',
      error: `Input length ${input.length} exceeds maximum ${MAX_INPUT_LENGTH} characters`,
    };
  }

  return process(input, config);
}

/**
 * Example: Streaming pattern for large inputs
 *
 * Use when input is too large to process all at once
 * Generator functions enable lazy evaluation
 */
function* processChunks(
  input: string,
  chunkSize: number,
  config: Config,
): Generator<Result, void, undefined> {
  for (let i = 0; i < input.length; i += chunkSize) {
    const chunk = input.slice(i, i + chunkSize);
    yield process(chunk, config);
  }
}

/**
 * Example: Async batch processing with concurrency limit
 *
 * Useful for I/O-bound operations
 */
async function processBatchAsync(
  inputs: string[],
  config: Config,
  concurrency: number = 5,
): Promise<Result[]> {
  const results = new Array<Result>(inputs.length);
  const executing: Promise<void>[] = [];

  for (let i = 0; i < inputs.length; i++) {
    const promise = Promise.resolve().then(() => {
      results[i] = process(inputs[i], config);
    });

    executing.push(promise);

    if (executing.length >= concurrency) {
      await Promise.race(executing);
      executing.splice(
        executing.findIndex((p) => p === promise),
        1,
      );
    }
  }

  await Promise.all(executing);
  return results;
}

/**
 * Example: Pre-sizing Sets and Maps
 *
 * Performance pattern for collections with known size
 */
function deduplicateWithPresize(inputs: string[]): string[] {
  // Note: JavaScript Set doesn't have capacity constructor
  // But we can pre-allocate the result array
  const unique = new Set(inputs);
  return Array.from(unique);
}

// Main execution
async function main() {
  console.log('Tool Name - Advanced Usage Patterns\n');

  const config: Config = { verbose: false };

  // Example 1: Batch processing with pre-allocation
  console.log('Example 1: Batch processing (optimized)');
  const inputs = ['hello', 'world', 'typescript', 'performance'];
  const results = batchProcessOptimized(inputs, config);
  console.log(`  Processed ${results.length} inputs:`);
  results.forEach((r, i) => {
    if (r.success) {
      console.log(`    [${i}] ${r.data}`);
    }
  });
  console.log();

  // Example 2: Resource limit handling
  console.log('Example 2: Resource limit');
  const largeInput = 'x'.repeat(2_000_000); // Exceeds MAX_INPUT_LENGTH
  const limitResult = processWithLimit(largeInput, config);
  if (!limitResult.success) {
    console.log(`  Error (expected): ${limitResult.error}`);
  }
  console.log();

  // Example 3: Streaming with generators
  console.log('Example 3: Streaming with generators');
  const largeText = 'hello world '.repeat(100); // 1200 chars
  let chunkCount = 0;
  for (const result of processChunks(largeText, 100, config)) {
    if (result.success) chunkCount++;
  }
  console.log(`  Processed ${chunkCount} chunks`);
  console.log();

  // Example 4: Async batch with concurrency limit
  console.log('Example 4: Async batch processing (concurrency: 3)');
  const asyncInputs = ['one', 'two', 'three', 'four', 'five'];
  const asyncResults = await processBatchAsync(asyncInputs, config, 3);
  console.log(`  Processed ${asyncResults.filter((r) => r.success).length} successfully`);
  console.log();

  // Example 5: Deduplication
  console.log('Example 5: Deduplication');
  const duplicates = ['hello', 'world', 'hello', 'typescript', 'world'];
  const unique = deduplicateWithPresize(duplicates);
  console.log(`  Input: ${duplicates.length} items, Unique: ${unique.length} items`);
  console.log();

  console.log('Performance Tips:');
  console.log('  - Pre-allocate arrays with new Array(size) when size is known');
  console.log('  - Check resource limits before processing large inputs');
  console.log('  - Use generators for lazy/streaming evaluation');
  console.log('  - Use async concurrency limits for I/O-bound operations');
  console.log('  - Profile with: node --prof your-script.js && node --prof-process isolate-*.log');
  console.log();

  console.log('Done!');
}

main().catch((err) => {
  console.error('Error:', err);
  process.exit(1);
});

File: templates/rust-tool-template/examples/advanced.rs
//! Advanced usage patterns for tool-name
//!
//! This example demonstrates:
//! - Performance optimization patterns
//! - Resource limit handling
//! - Batch processing with pre-allocation
//! - Error handling strategies
//!
//! Run this example:
//!   cargo run --example advanced

use tool_name::{process, Config, ProcessError};

/// Example: Pre-allocate collections when size is known
///
/// Performance pattern used in output-diffing-utility:
/// - Vec::with_capacity() avoids repeated reallocation
/// - Reduces memory fragmentation
/// - Improves performance for known-size collections
fn batch_process_optimized(inputs: &[&str], config: &Config) -> Result<Vec<String>, ProcessError> {
    // Pre-allocate: we know we'll have exactly inputs.len() results
    let mut results = Vec::with_capacity(inputs.len());

    for input in inputs {
        results.push(process(input, config)?);
    }

    Ok(results)
}

/// Example: Resource limit pattern
///
/// Pattern used in output-diffing-utility for file size limits:
/// - Define constants for resource limits
/// - Check limits before processing
/// - Provide clear error messages with usage hints
const MAX_INPUT_LENGTH: usize = 1_000_000; // 1 million characters

fn process_with_limit(input: &str, config: &Config) -> Result<String, ProcessError> {
    // Check resource limit before processing
    if input.len() > MAX_INPUT_LENGTH {
        return Err(ProcessError::InvalidInput(
            format!("Input length {} exceeds maximum {} characters",
                    input.len(), MAX_INPUT_LENGTH)
        ));
    }

    process(input, config)
}

/// Example: Streaming/chunked processing for large inputs
///
/// Use when input is too large to process all at once
fn process_chunks(input: &str, chunk_size: usize, config: &Config) -> Result<Vec<String>, ProcessError> {
    let chunks: Vec<&str> = input
        .as_bytes()
        .chunks(chunk_size)
        .filter_map(|chunk| std::str::from_utf8(chunk).ok())
        .collect();

    // Pre-allocate for all chunks
    let mut results = Vec::with_capacity(chunks.len());

    for chunk in chunks {
        results.push(process(chunk, config)?);
    }

    Ok(results)
}

/// Example: Iterator-based processing (no intermediate collections)
///
/// More memory-efficient than collecting all results
fn process_iterator<'a>(
    inputs: &'a [&'a str],
    config: &'a Config,
) -> impl Iterator<Item = Result<String, ProcessError>> + 'a {
    inputs.iter().map(move |input| process(input, config))
}

fn main() {
    println!("Tool Name - Advanced Usage Patterns\n");

    let config = Config::default();

    // Example 1: Batch processing with pre-allocation
    println!("Example 1: Batch processing (optimized)");
    let inputs = vec!["hello", "world", "rust", "performance"];
    match batch_process_optimized(&inputs, &config) {
        Ok(results) => {
            println!("  Processed {} inputs:", results.len());
            for (i, result) in results.iter().enumerate() {
                println!("    [{}] {}", i, result);
            }
        }
        Err(e) => eprintln!("  Error: {}", e),
    }
    println!();

    // Example 2: Resource limit handling
    println!("Example 2: Resource limit");
    let large_input = "x".repeat(2_000_000); // Exceeds MAX_INPUT_LENGTH
    match process_with_limit(&large_input, &config) {
        Ok(_) => println!("  Unexpected success"),
        Err(e) => println!("  Error (expected): {}", e),
    }
    println!();

    // Example 3: Chunked processing
    println!("Example 3: Chunked processing");
    let large_text = "hello world ".repeat(100); // 1200 chars
    match process_chunks(&large_text, 100, &config) {
        Ok(chunks) => println!("  Processed {} chunks", chunks.len()),
        Err(e) => eprintln!("  Error: {}", e),
    }
    println!();

    // Example 4: Iterator-based processing (lazy evaluation)
    println!("Example 4: Iterator processing (lazy)");
    let inputs = vec!["one", "two", "three"];
    let mut success_count = 0;
    let mut error_count = 0;

    for result in process_iterator(&inputs, &config) {
        match result {
            Ok(_) => success_count += 1,
            Err(_) => error_count += 1,
        }
    }
    println!("  Successes: {}, Errors: {}", success_count, error_count);
    println!();

    // Example 5: Collecting only successful results
    println!("Example 5: Filter successful results");
    let mixed_inputs = vec!["valid", "", "also valid"];
    let results: Vec<String> = process_iterator(&mixed_inputs, &config)
        .filter_map(Result::ok)
        .collect();
    println!("  Valid results: {}", results.len());
    println!();

    println!("Performance Tips:");
    println!("  - Use Vec::with_capacity() when final size is known");
    println!("  - Check resource limits before processing");
    println!("  - Use iterators for lazy evaluation");
    println!("  - Profile with: cargo build --release && time ./target/release/tool-name");
    println!();

    println!("Done!");
}
